certora/specs/LiquidityPoolMoneyFlow.spec

```diff
@@ -41,111 +51,186 @@ function CVLCallHook(env e)  {
     } else if (nondeterministic_choice == 2) {
-        satisfy money_recipient == bnftHolder;
+        satisfy sent_bnftHolder;
     } else {
-        satisfy money_recipient == tnftHolder;
+        satisfy sent_tnftHolder;
     }
 
 }
 
+// Money only flows to the expected recipients for partial withdraw
+/*
 rule money_flow_from_node_partial_withdraw {
     uint256 validatorId;
     env e;
 
-    // initialize ghost to currentContract
-    require money_recipient == currentContract; 
+    // // expected party addresses
+    require treasury == currentContract.treasuryContract;
+    require nodeOperator == auctionManager.getBidOwner(e, validatorId);
+    require bnftHolder == currentContract.bnft.ownerOf(e, validatorId);
+    require tnftHolder == currentContract.tnft.ownerOf(e, validatorId);
 
-    // expected party addresses
-    address treasury = currentContract.treasuryContract;
-    address nodeOperator = auctionManager.getBidOwner(e, validatorId);
-    address bnftHolder = currentContract.bnft.ownerOf(e, validatorId);
-    address tnftHolder = currentContract.tnft.ownerOf(e, validatorId);
+    require !illegal_transfer;
+
+    require !sent_treasury;
+    require !sent_nodeOperator;
+    require !sent_bnftHolder;
+    require !sent_tnftHolder;
 
     partialWithdraw(e, validatorId);
 
-    // If money was moved during this call it goes
-    // to one of the expected recipients. The
-    // case where money_recipient is currentContract
-    // models the case where nonzero money is not moved.
-    assert money_recipient == currentContract ||
-        money_recipient == treasury ||
-        money_recipient == nodeOperator ||
-        money_recipient == bnftHolder ||
-        money_recipient == tnftHolder; 
+    assert !illegal_transfer;
 
     // The following encodes that all of these recipient
     // cases are reachable (and ensures that the above
     // assertion isn't just trivially stuck in the initial state)
 
+    satisfy treasury != nodeOperator &&
+        treasury != bnftHolder &&
+        treasury != tnftHolder &&
+        nodeOperator != bnftHolder &&
+        nodeOperator != tnftHolder &&
+        bnftHolder != tnftHolder;
+
 
     // This unconstrained ghost on which we branch is
     // just here to encode nondeterminstic choice
     mathint nondeterministic_choice;
+
     if (nondeterministic_choice == 0) {
-        satisfy money_recipient == treasury;
+        satisfy sent_treasury;
     } else if (nondeterministic_choice == 1) {
-        satisfy money_recipient == nodeOperator;
+        satisfy sent_nodeOperator;
     } else if (nondeterministic_choice == 2) {
-        satisfy money_recipient == bnftHolder;
+        satisfy sent_bnftHolder;
     } else {
-        satisfy money_recipient == tnftHolder;
+        satisfy sent_tnftHolder;
     }
 }
-
+*/
+
+// These cause hardstops
+// // frontrunning cannot cause withdraw to be blocked
+// rule money_flow_from_node_full_withdraw_frontrunning (method f) {
```

Comment:
Yes, that's expected. It's a difficult rule.
Maybe we can start by filtering to a single method that also involves ETH transfer?

Reply:
Just by simulating this in my head (and also from past experience with similar rules),
The second call could revert if the contract doesn't have enough native balance to satisfy both transfers (that is, the original and the front-runner).
So it has enough ETH to satisfy each of the transfers individually, and not together.

This naturally leads to a need for some solvency assumption - that the contract has enough ETH for two validators at the same time.
We might deviate a bit from the original rule by trying to prove it (which isn't trivial) so perhaps we could simply assume it.